### PR TITLE
console elect for cross dc leader

### PR DESCRIPTION
**Pull request by LanternLee**

### CODE REVIEW

**Comment by wenchaomeng** [comment]
单测覆盖率就要跌破 55% 了

File: redis/redis-console/src/main/java/com/ctrip/xpipe/redis/console/model/LeaseModel.java [hunk 1]
```diff
@@ -0,0 +1,39 @@
+package com.ctrip.xpipe.redis.console.model;
+
+public class LeaseModel {
+
+    private String name;
+
+    private String owner;
+
+    private Integer validPeriod;
```
Comment (NickNYU) [inline redis/redis-console/src/main/java/com/ctrip/xpipe/redis/console/model/LeaseModel.java:9]: long?

File: redis/redis-console/src/main/java/com/ctrip/xpipe/redis/console/controller/api/CrossDcController.java [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+    @Autowired(required = false)
+    private ConsoleCrossDcServer crossDcClusterServer;
+
+    @Autowired
+    private DcService dcService;
+
+    private final static int DEFAULT_LEASE_PERIOD = 10;
+
+    @RequestMapping(value = "/cross-dc/lease", method = RequestMethod.POST)
+    public RetMessage updateCrossDcLeaderLease(HttpServletRequest request, @RequestBody LeaseModel leaseModel) {
```
Comment (NickNYU) [inline redis/redis-console/src/main/java/com/ctrip/xpipe/redis/console/controller/api/CrossDcController.java:32]: 不用 request body

Comment (NickNYU) [inline redis/redis-console/src/main/java/com/ctrip/xpipe/redis/console/controller/api/CrossDcController.java:32]: curl -X POST http://xpipe.ctripcorp.com/cross-dc/lease/SHAJQ/10

File: redis/redis-console/src/main/java/com/ctrip/xpipe/redis/console/election/CrossDcLeaderElectionAction.java [hunk 3]
```diff
@@ -0,0 +1,192 @@
...
+
+    protected String localIp = FoundationService.DEFAULT.getLocalIp();
+
+    @Autowired
+    private ConfigDao configDao;
+
+    @Autowired
+    private MetaCache metaCache;
+
+    private ConfigTbl currentConfig;
```
Comment (NickNYU) [inline redis/redis-console/src/main/java/com/ctrip/xpipe/redis/console/election/CrossDcLeaderElectionAction.java:42]: lease

File: redis/redis-console/src/main/java/com/ctrip/xpipe/redis/console/election/CrossDcLeaderElectionAction.java [hunk 4]
```diff
@@ -0,0 +1,192 @@
...
+
+        if (0 == totalCluster) return 0;
+        return activeClusterCount / (totalCluster * 1f);
+    }
+
+    private boolean isConfigExpired() {
+        return null != currentConfig && (new Date()).compareTo(currentConfig.getUntil()) >= 0;
+    }
+
+    private boolean isConfigActive() {
```
Comment (NickNYU) [inline redis/redis-console/src/main/java/com/ctrip/xpipe/redis/console/election/CrossDcLeaderElectionAction.java:170]: isOnLease

File: redis/redis-console/src/main/java/com/ctrip/xpipe/redis/console/election/CrossDcLeaderElectionAction.java [hunk 5]
```diff
@@ -0,0 +1,192 @@
...
+                activeClusterCount += count;
+            }
+            totalCluster += count;
+        }
+
+        if (0 == totalCluster) return 0;
+        return activeClusterCount / (totalCluster * 1f);
+    }
+
+    private boolean isConfigExpired() {
```
Comment (NickNYU) [inline redis/redis-console/src/main/java/com/ctrip/xpipe/redis/console/election/CrossDcLeaderElectionAction.java:166]: isLeaseExpired

File: redis/redis-console/src/main/java/com/ctrip/xpipe/redis/console/election/CrossDcLeaderElectionAction.java [hunk 6]
```diff
@@ -0,0 +1,192 @@
...
+        config.setKey(KEY_LEASE_CONFIG)
+                .setSubKey(SUB_KEY_CROSS_DC_LEADER)
+                .setVal(dataCenter)
+                .setUpdateIP(localIp)
+                .setUpdateUser(dataCenter + "-DcLeader");
+
+        return config;
+    }
+
+    private long calculateElectDelay() {
```
Comment (NickNYU) [inline redis/redis-console/src/main/java/com/ctrip/xpipe/redis/console/election/CrossDcLeaderElectionAction.java:136]: 注释

**Comment by NickNYU** [comment]
> 单测覆盖率就要跌破 55% 了

目前已解决, 之前是 REST Controller 使用不对, 导致有些单测跑不到

File: redis/redis-console/src/main/java/com/ctrip/xpipe/redis/console/controller/api/CrossDcController.java (hunk 2 above)
Comment (LanternLee) [reply redis/redis-console/src/main/java/com/ctrip/xpipe/redis/console/controller/api/CrossDcController.java]: curl -X POST  http://xpipe.ctripcorp.com/api/cross-dc/leader/force

**Pull request merged**